solutions/sapo-14-genghis.mdx

```diff
@@ -0,0 +1,240 @@
+		for (int start = 0; start + num - 1 < stable_num + 2; start++) {
+			int end = start + num - 1;
+			/*
+			 * first handle the beginning and ending edge cases,
+			 * where the smallest or largest element has become the root
+			 */
+			lowest_ops[start][end] = std::min(
+				lowest_ops[start][end],
+				lowest_ops[start][start + 1]
+				+ lowest_ops[start + 1][end]
+				+ stable_horses[end] - stable_horses[start + 1]
+			);
+			lowest_ops[start][end] = std::min(
+				lowest_ops[start][end],
+				lowest_ops[end - 1][end]
+				+ lowest_ops[start][end - 1]
+				+ stable_horses[end - 1] - stable_horses[start]
+			);
+
+			/*
+			 * then get to the actual part where we combine
+			 * two parts with top being the root
+			 */
+			for (int top = start + 2; top < end - 1; top++) {
+				lowest_ops[start][end] = std::min(
+					lowest_ops[start][end],
+					lowest_ops[start][top]
+					+ lowest_ops[top][end]
+					+ stable_horses[end] - stable_horses[start] - 1
+				);
+			}
+		}
+	}
+	cout << lowest_ops[0][stable_num + 1] << endl;
+}
+```
+
+</CPPSection>
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.Arrays;
+
+public final class genghis {
+	public static void main(String[] args) throws IOException {
+		BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
+		int horseNum = Integer.parseInt(read.readLine());
+		int stableNum = Integer.parseInt(read.readLine());
+		int[] stableHorses = new int[stableNum + 2];
+		for (int i = 1; i <= stableNum; i++) {
+			stableHorses[i] = Integer.parseInt(read.readLine());
+			if (stableHorses[i] < 1 || horseNum < stableHorses[i]) {
+				throw new IllegalArgumentException("this horse shouldn't exist");
+			}
+		}
+		// add the "out-of-bounds" horses
+		stableHorses[0] = 0;
+		stableHorses[stableNum + 1] = horseNum + 1;
+		Arrays.sort(stableHorses);
+
+		/*
+ 		 * this[i][j] = min operations if we only consider horses
+		 * from that exist in the index interval (i, j)
+		 * ex: this[1][4] and stable_horses = [0, 1, 3, 7, 10, 11]
+		 *     means that we only consider horses from (1, 10)
+		 */
+		int[][] lowestOps = new int[stableNum + 2][stableNum + 2];
+		for (int i = 0; i < stableNum + 2; i++) {
+			Arrays.fill(lowestOps[i], Integer.MAX_VALUE);
+			lowestOps[i][i] = 0;
+			if (i + 1 < stableNum + 2) {
+				lowestOps[i][i + 1] = stableHorses[i + 1] - stableHorses[i] - 1;
+			}
+			if (i + 2 < stableNum + 2) {
+				lowestOps[i][i + 2] = stableHorses[i + 2] - stableHorses[i] - 1;
+			}
+		}
+		for (int num = 4; num <= stableNum + 2; num++) {
+			for (int start = 0; start + num - 1 < stableNum + 2; start++) {
+				int end =  start + num - 1;
+				/*
+				 * first handle the beginning and ending edge cases,
+				 * where the smallest or largest element has become the root
+				 */
+				lowestOps[start][end] = Math.min(
+						lowestOps[start][end],
+						lowestOps[start][start + 1]
+								+ lowestOps[start + 1][end]
+								+ stableHorses[end] - stableHorses[start + 1]
```

Comment:
why is this indented here but not in the cpp or python code?

Reply:
ok so for python intellij said i shouldn't indent
but for java it said i should
and for c++ i usually stick to the python guidelines

Reply:
didn't get a response so just gonna